components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/AbstractLocalApplicationAuthenticator.java

```diff
@@ -0,0 +1,289 @@
+                    context.setProperty(FrameworkConstants.LAST_FAILED_AUTHENTICATOR, null);
+                    fireEvent(context, IdentityEventConstants.Event.POST_AUTHENTICATION, true);
+                    return AuthenticatorFlowStatus.SUCCESS_COMPLETED;
+                } catch (AuthenticationFailedException e) {
+                    if (checkUserAccountStatus(response, context)) {
+                        return AuthenticatorFlowStatus.INCOMPLETE;
+                    }
+                    fireEvent(context, IdentityEventConstants.Event.POST_AUTHENTICATION, false);
+                    request.setAttribute(FrameworkConstants.REQ_ATTR_HANDLED, true);
+                    // Decide whether we need to redirect to the login page to retry authentication.
+                    return handleRetryOnFailure(request, response, context, e);
+                }
+            }
+            // if a logout flow
+        } else {
+            return processLogoutFlow(request, response, context);
+        }
+    }
+
+    /**
+     * To decide whether need to redirect the user to login page to retry authentication.
+     *
+     * @param request  the httpServletRequest
+     * @param response the httpServletResponse
+     * @param context  the authentication context
+     * @param e        the authentication failed exception
+     * @return authentication flow status
+     * @throws AuthenticationFailedException the exception in the authentication flow
+     */
+    protected AuthenticatorFlowStatus handleRetryOnFailure(HttpServletRequest request,
+                                                                    HttpServletResponse response,
+                                                                    AuthenticationContext context,
+                                                                    AuthenticationFailedException e)
+            throws AuthenticationFailedException {
+
+        boolean sendToMultiOptionPage =
+                isStepHasMultiOption(context) && isRedirectToMultiOptionPageOnFailure();
+        if (retryAuthenticationEnabled(context) && !sendToMultiOptionPage) {
+            // The Authenticator will re-initiate the authentication and retry.
+            context.setRetrying(true);
+            return initiateAuthenticationFlow(request, response, context);
+        } else {
+            context.setProperty(FrameworkConstants.LAST_FAILED_AUTHENTICATOR, getName());
+            // By throwing this exception step handler will redirect to multi options page if
+            // multi-option are available in the step.
+            throw e;
+        }
+    }
+
+    /**
+     * To check whether user domain and tenant domain equal for non SaaS application.
+     *
+     * @param context the authentication context
+     * @throws AuthenticationFailedException the exception in the authentication flow
+     */
+    protected void processWithNonSaaSApp(AuthenticationContext context) throws AuthenticationFailedException {
```

Comment:
can we have it like 'validateNonSaasAppLogin()'? The name here is not understandable